Allow to dynamically add strategy in the plugin init method (#1479)
kuzzleio/kuzzle: doc/1/plugins/guides/strategies/overview/index.md

};
```

## Register a new strategy

There are two ways of registering new strategies:

- statically, by exposing a `strategies` object
- dynamically, by using the dedicated [strategy accessors](/core/1/plugins/plugin-context/accessors/strategies)

Whether strategies are added statically or dynamically, the registered strategy must expose the following properties:

| Arguments | Type              | Description                           |
| --------- | ----------------- | ------------------------------------- |
| `config`  | <pre>object</pre> | Authentication strategy configuration |
| `methods` | <pre>object</pre> | List of exposed methods               |

### statically register strategies

> Plugins can declare a `strategies` object which contain authentication strategies to register.  

Reviewer: typo + suggestion:

```suggestion
Plugins can declare a `strategies` object which contains the authentication strategies to register.  
This object will be interpreted by Kuzzle only once, immediately after this plugin's init function has resolved.  
```

Each key of this object is the name of the strategy to register and the value is the strategy object containing `config` and `methods` properties.  

For example, to register a strategy named `local` with the `Local` authenticator:
```js
this.authenticators = {
  Local: require('passport-local')
};

this.strategies = {
  local: {
    config: {
      authenticator: 'Local'
    },

>     // theses methods must exists in the plugin object

Reviewer: ```suggestion
    // these methods must be exposed by the plugin 
```

    methods: {
      create: 'create',
      delete: 'delete',
      exists: 'exists',
      getById: 'getById',
      getInfo: 'getInfo',
      update: 'update',
      validate: 'validate',
      verify: 'verify'
    }
  }
}
```

### dynamically register strategies

Strategies can be register at runtime with the [strategies.add](/core/1/plugins/plugin-context/accessors/strategies/#add) method.  

::: info

> Strategies added dynamically in the plugin [init method](/core/1/plugins/guides/manual-setup/init-function/) are added to the static `strategies` object and loaded by Kuzzle after plugin initialization.

Reviewer: ```suggestion
Strategies added dynamically in the plugin's [init method](/core/1/plugins/guides/manual-setup/init-function/) are added to the static `strategies` object and loaded by Kuzzle after the plugin initialization.
```
